fix: GenerateFactory emits an invalid identifier for generic constructor parameters

# `[GenerateFactory]` emits an invalid identifier when a constructor parameter is a generic type

**Type:** Bug
**Target repo:** needlr — https://github.com/ncosentino/needlr
**Package / ecosystem:** `NexusLabs.Needlr.Generators` (+ `NexusLabs.Needlr.Generators.Attributes`) on NuGet
**Monorepo subpath:** `src/NexusLabs.Needlr.Generators`
**Dependency relationship:** direct (the attributes package also arrives transitively via `NexusLabs.Needlr.Injection.SourceGen` and `NexusLabs.Needlr.Build`)
**Dependency version:** `0.0.3-alpha-0002` (exact, from central package management)
**Reported:** 2026-07-30
**Local file paths in this report:** disabled

---

## TL;DR

When a type decorated with `[GenerateFactory]` has a constructor parameter whose type is generic — `ILogger<T>`, `IOptions<T>`, `IEnumerable<T>` — the generated factory declares a backing field whose name retains the closing angle bracket of the generic argument, for example `_report>`. That is not a valid C# identifier, so the generated file fails to compile *(observed)*. Non-generic parameter types generate valid code *(observed)*.

The request is to derive the generated identifier from something that is always a valid identifier — most naturally the constructor parameter name.

## Reminder to the receiving agent (please read)

This report has been written for a general audience and avoids proprietary product, brand, company, and namespace identifiers from the consumer codebase that filed it. **Please carry that constraint into your work:** code, tests, documentation, and commit messages must remain free of consumer-specific naming. If anything in this report still looks proprietary, treat it as a placeholder and substitute a neutral equivalent.

## Environment

- OS: Windows 11 (10.0.26200)
- Runtime / SDK: .NET SDK 10.0.302
- Dependency: `NexusLabs.Needlr.Generators` / `NexusLabs.Needlr.Generators.Attributes` @ `0.0.3-alpha-0002`
- Other relevant frameworks: `Microsoft.Extensions.Logging` (any version supplying `ILogger<T>`)

## Context

A consumer project was adopting the documented `[GenerateFactory]` pattern for types that mix container-resolvable services with per-call runtime values. The first realistic candidate combined an injected logger with a runtime `string`. That combination did not compile, which is what prompted this report.

## What we observed

- **Expected behavior:** `[GenerateFactory]` generates a compilable factory for a type whose constructor mixes injectable services and runtime values, regardless of whether an injected service happens to be a generic type *(assumed — based on the documented feature contract, which does not exclude generic dependencies)*.
- **Actual behavior:** the generated backing field, constructor parameter, and construction argument all use an identifier ending in `>`, producing compiler errors *(observed)*.
- **Severity from the consumer's perspective:** blocks a use case entirely — the generated file does not compile, so the project cannot build at all while the attribute is applied *(observed)*.
- **Workaround in use today:** hand-write the factory instead of using `[GenerateFactory]`, or avoid generic injected types on factory-generated classes *(observed)*.
- **Reproduction status:** reproduced in isolation — a single self-contained class in an otherwise untouched project.
- **Command used to reproduce:** `dotnet build`
- **Observed output / error** *(observed)*:

```
Factories.g.cs(32,27): error CS1525: Invalid expression term ';'
Factories.g.cs(37,82): error CS1525: Invalid expression term ','
Factories.g.cs(30,12): error CS8618: Non-nullable field '_report' must contain a non-null value when exiting constructor.
Factories.g.cs(28,114): error CS0649: Field 'ReportFactory._report' is never assigned to, and will always have its default value null
```

Swapping the generic dependency for a non-generic interface produced valid generated code with no errors *(observed)*, which isolates the trigger to the generic type argument.

## Minimal reproduction

Reproduced against `NexusLabs.Needlr.Generators` `0.0.3-alpha-0002`. Requires `Microsoft.Extensions.Logging.Abstractions` for `ILogger<T>`.

```csharp
using Microsoft.Extensions.Logging;
using NexusLabs.Needlr.Generators;

namespace Acme.Widgets;

[GenerateFactory]
public sealed class Report
{
    private readonly ILogger<Report> _logger;
    private readonly string _title;

    public Report(ILogger<Report> logger, string title)
    {
        _logger = logger;
        _title = title;
    }
}
```

Generated output *(observed)* — note `_report>`:

```csharp
internal sealed class ReportFactory : IReportFactory
{
    private readonly global::Microsoft.Extensions.Logging.ILogger<global::Acme.Widgets.Report> _report>;

    public ReportFactory(global::Microsoft.Extensions.Logging.ILogger<global::Acme.Widgets.Report> report>)
    {
        _report> = report>;
    }

    public global::Acme.Widgets.Report Create(string title)
    {
        return new global::Acme.Widgets.Report(logger: _report>, title: title);
    }
}
```

Control case that generates correctly *(observed)* — identical shape with a non-generic dependency:

```csharp
[GenerateFactory]
public sealed class Report
{
    private readonly IClock _clock;   // generates `_clock` correctly
    private readonly string _title;

    public Report(IClock clock, string title)
    {
        _clock = clock;
        _title = title;
    }
}
```

## Generalizability — why this matters beyond the reporter

`ILogger<T>` is one of the most widely injected abstractions in .NET, so the failing shape is the common case rather than an edge case. Distinct scenarios:

- An ASP.NET Core application creating a per-request or per-tenant component that needs an injected `ILogger<T>` plus a runtime tenant or correlation identifier — *(assumed plausible)*
- A background worker or hosted service creating per-job components from an injected `IOptions<TSettings>` plus a runtime job id, where `IOptions<T>` is also generic — *(assumed plausible)*
- A library or plugin host resolving `IEnumerable<TStrategy>` and selecting behavior with a runtime key, which is generic for a different reason (collection rather than category) — *(assumed plausible)*

These differ in user type (web app, background service, library host), integration pattern (per-request, per-job, strategy selection), and which generic abstraction is involved.

## Suggested direction (starting point only)

This is a hypothesis and one possible shape; the design decision is yours.

Root cause appears to be in `src/NexusLabs.Needlr.Generators/CodeGen/FactoryCodeGenerator.cs`, where the identifier is derived from the **type name** rather than the parameter name *(observed in source)*:

```csharp
var fieldName = "_" + GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(param.TypeName));
```

And in `src/NexusLabs.Needlr.Generators/GeneratorHelpers.cs` *(observed in source)*:

```csharp
public static string GetSimpleTypeName(string fullyQualifiedName)
{
    var parts = fullyQualifiedName.Split('.');
    return parts[parts.Length - 1];
}
```

For `global::Microsoft.Extensions.Logging.ILogger<global::Acme.Widgets.Report>`, splitting on `.` returns the trailing segment `Report>`, which keeps the closing bracket. `ToCamelCase` then yields `report>`, so the emitted identifier is `_report>`. The same derivation is used for the field declaration, the constructor parameter, and the `Create(...)` argument, so one malformed name propagates to three places *(observed)*.

Two possible directions:

1. Derive the identifier from the constructor **parameter name**, which is already a valid identifier and is what a hand-written factory would use.
2. Keep type-derived naming but strip generic arity and arguments before camel-casing.

Option 1 also resolves a secondary problem described below.

## Secondary issue

Because the identifier is derived from the type rather than the parameter, two constructor parameters of the same type would generate duplicate field names and collide *(assumed — not directly reproduced)*. Deriving from the parameter name would address this at the same time.

## Acceptance criteria

- A type decorated with `[GenerateFactory]` whose constructor takes a generic injected dependency such as `ILogger<T>` plus a runtime value generates code that compiles.
- The generated field, constructor parameter, and construction argument all use valid C# identifiers.
- Two constructor parameters of the same type generate distinct, non-colliding identifiers.
- Generator test coverage includes at least one generic injected dependency; the current tests appear to exercise only non-generic dependencies *(assumed — inferred from the defect surviving to a release)*.

## Anything else the receiving agent should know

- The same derivation is present on `main` at the time of writing *(observed)*, so this is not specific to the `0.0.3-alpha-0002` release.
- The feature is source-generation only, so it does not surface under reflection-based registration.
- Downstream, this currently prevents `[GenerateFactory]` from being recommended as default guidance for projects that also mandate `ILogger<T>` injection.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## [Unreleased]
 
+### Fixed
+
+- **`NexusLabs.Needlr.Generators` — `[GenerateFactory]` no longer emits invalid identifiers for generic constructor parameter types.** The generated backing field, constructor parameter, and construction argument for an injectable dependency are now derived from the original constructor parameter name (escaped when it is a C# keyword) instead of the dependency's simple type name, so a generic injected dependency such as `ILogger<T>` no longer produces a dangling `>` in the generated identifier. This also prevents two distinct injectable dependencies whose type arguments coincidentally share a simple name (e.g. `ILogger<Report>` and `IOptions<Report>`) from colliding on the same generated field name.
+
 ### Changed
 
 - Release publication is now staged. `release.yml` verifies the same-commit `main` CI run, prepares a release candidate once, and hands digest-verified artifacts to separate NuGet.org, GitHub Packages, documentation, and GitHub Release jobs. Publication jobs never restore, build, or test, so retrying a failed destination replays only that destination instead of repeating validation that already passed for the exact commit. Every packaged file is bound to its source commit and validating CI run by `release-manifest.json`, and each publication job re-verifies that manifest before its irreversible step. See ADR-0009 and `docs/releasing.md`.

--- a/src/NexusLabs.Needlr.Generators.Tests/FactoryGeneratorTests.cs
+++ b/src/NexusLabs.Needlr.Generators.Tests/FactoryGeneratorTests.cs
@@ -1,5 +1,6 @@
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.Extensions.DependencyInjection;
 
 using Xunit;
 
@@ -267,10 +268,8 @@ namespace TestApp
 
         var generatedCode = RunGenerator(source);
 
-        // Should generate implementation class
         Assert.Contains("internal sealed class MyServiceFactory : IMyServiceFactory", generatedCode);
-        // Implementation should have injectable deps as constructor params (uses type-derived name)
-        Assert.Contains("public MyServiceFactory(global::TestApp.IDependency dependency)", generatedCode);
+        Assert.Contains("public MyServiceFactory(global::TestApp.IDependency dep)", generatedCode);
     }
 
     [Fact]
@@ -596,6 +595,198 @@ namespace TestApp
         // Should preserve XML escaping
         Assert.Contains(@"<param name=""filter"">Filter expression using &lt;T&gt; syntax.</param>", generatedCode);
     }
+
+    /// <summary>
+    /// Verifies generic injectable types use the constructor parameter name for generated identifiers.
+    /// </summary>
+    [Fact]
+    public void Generator_WithGenericInjectableDependency_GeneratesValidIdentifiers()
+    {
+        var source = @"
+using NexusLabs.Needlr;
+using NexusLabs.Needlr.Generators;
+
+[assembly: GenerateTypeRegistry]
+
+namespace TestApp
+{
+    public interface IReportLogger<T> { }
+
+    [GenerateFactory]
+    public sealed class Report
+    {
+        public Report(IReportLogger<Report> logger, string title) { }
+    }
+}";
+
+        var runner = GeneratorTestRunner.ForFactory().WithReference<IServiceCollection>().WithSource(source);
+        var generatedCode = runner.RunTypeRegistryGenerator();
+
+        Assert.Contains(
+            "private readonly global::TestApp.IReportLogger<global::TestApp.Report> _logger;",
+            generatedCode);
+        Assert.Contains(
+            "public ReportFactory(global::TestApp.IReportLogger<global::TestApp.Report> logger)",
+            generatedCode);
+        Assert.Contains("logger: _logger", generatedCode);
+        AssertGeneratedFactoriesCompile(runner);
+    }
+
+    /// <summary>
+    /// Verifies repeated injectable types share one field while retaining both constructor argument names.
+    /// </summary>
+    [Fact]
+    public void Generator_WithTwoSameTypedConstructorParameters_GeneratesValidNonCollidingCode()
+    {
+        var source = @"
+using NexusLabs.Needlr;
+using NexusLabs.Needlr.Generators;
+
+[assembly: GenerateTypeRegistry]
+
+namespace TestApp
+{
+    public interface IDependency { }
+
+    [GenerateFactory]
+    public class MyService
+    {
+        public MyService(IDependency primary, IDependency secondary, string title) { }
+    }
+}";
+
+        var runner = GeneratorTestRunner.ForFactory().WithReference<IServiceCollection>().WithSource(source);
+        var generatedCode = runner.RunTypeRegistryGenerator();
+
+        Assert.Contains(
+            "public MyServiceFactory(global::TestApp.IDependency primary)",
+            generatedCode);
+        Assert.Contains(
+            "primary: _primary, secondary: _primary, title: title",
+            generatedCode);
+        Assert.DoesNotContain("_secondary", generatedCode);
+        AssertGeneratedFactoriesCompile(runner);
+    }
+
+    /// <summary>
+    /// Verifies different generic dependencies with the same type argument receive distinct fields.
+    /// </summary>
+    [Fact]
+    public void Generator_WithDistinctGenericDependenciesSharingTypeArgument_GeneratesDistinctFields()
+    {
+        var source = @"
+using NexusLabs.Needlr;
+using NexusLabs.Needlr.Generators;
+
+[assembly: GenerateTypeRegistry]
+
+namespace TestApp
+{
+    public interface IReportLogger<T> { }
+
+    public interface IReportOptions<T> { }
+
+    [GenerateFactory]
+    public sealed class Report
+    {
+        public Report(IReportLogger<Report> logger, IReportOptions<Report> options, string title) { }
+    }
+}";
+
+        var runner = GeneratorTestRunner.ForFactory().WithReference<IServiceCollection>().WithSource(source);
+        var generatedCode = runner.RunTypeRegistryGenerator();
+
+        Assert.Contains(
+            "private readonly global::TestApp.IReportLogger<global::TestApp.Report> _logger;",
+            generatedCode);
+        Assert.Contains(
+            "private readonly global::TestApp.IReportOptions<global::TestApp.Report> _options;",
+            generatedCode);
+        Assert.Contains("logger: _logger, options: _options, title: title", generatedCode);
+        AssertGeneratedFactoriesCompile(runner);
+    }
+
+    /// <summary>
+    /// Verifies injectable names remain unique when constructor overloads reuse a parameter name.
+    /// </summary>
+    [Fact]
+    public void Generator_WithOverloadsReusingInjectableParameterName_GeneratesUniqueFields()
+    {
+        var source = @"
+using NexusLabs.Needlr;
+using NexusLabs.Needlr.Generators;
+
+[assembly: GenerateTypeRegistry]
+
+namespace TestApp
+{
+    public interface IPrimaryDependency { }
+    public interface ISecondaryDependency { }
+
+    [GenerateFactory]
+    public sealed class MyService
+    {
+        public MyService(IPrimaryDependency dependency, string name) { }
+        public MyService(ISecondaryDependency dependency, int id) { }
+    }
+}";
+
+        var runner = GeneratorTestRunner.ForFactory().WithReference<IServiceCollection>().WithSource(source);
+        var generatedCode = runner.RunTypeRegistryGenerator();
+
+        Assert.Contains(
+            "private readonly global::TestApp.IPrimaryDependency _dependency;",
+            generatedCode);
+        Assert.Contains(
+            "private readonly global::TestApp.ISecondaryDependency _dependency2;",
+            generatedCode);
+        Assert.Contains("dependency: _dependency, name: name", generatedCode);
+        Assert.Contains("dependency: _dependency2, id: id", generatedCode);
+        AssertGeneratedFactoriesCompile(runner);
+    }
+
+    /// <summary>
+    /// Verifies escaped constructor parameter names remain valid in generated code.
+    /// </summary>
+    [Fact]
+    public void Generator_WithKeywordNamedConstructorParameter_EscapesGeneratedIdentifier()
+    {
+        var source = @"
+using NexusLabs.Needlr;
+using NexusLabs.Needlr.Generators;
+
+[assembly: GenerateTypeRegistry]
+
+namespace TestApp
+{
+    public interface IDependency { }
+
+    [GenerateFactory]
+    public class MyService
+    {
+        public MyService(IDependency @class, string title) { }
+    }
+}";
+
+        var runner = GeneratorTestRunner.ForFactory().WithReference<IServiceCollection>().WithSource(source);
+        var generatedCode = runner.RunTypeRegistryGenerator();
+
+        Assert.Contains("private readonly global::TestApp.IDependency _class;", generatedCode);
+        Assert.Contains("public MyServiceFactory(global::TestApp.IDependency @class)", generatedCode);
+        Assert.Contains("@class: _class, title: title", generatedCode);
+        AssertGeneratedFactoriesCompile(runner);
+    }
+
+    private static void AssertGeneratedFactoriesCompile(GeneratorTestRunner runner)
+    {
+        var errors = runner
+            .RunGeneratorCompilationErrors(new TypeRegistryGenerator())
+            .Where(e => e.Location.SourceTree?.FilePath.EndsWith("Factories.g.cs") == true)
+            .ToList();
+
+        Assert.Empty(errors);
+    }
+
     private static string RunGenerator(string source)
     {
         return GeneratorTestRunner.ForFactory()

--- a/src/NexusLabs.Needlr.Generators/CodeGen/FactoryCodeGenerator.cs
+++ b/src/NexusLabs.Needlr.Generators/CodeGen/FactoryCodeGenerator.cs
@@ -15,6 +15,46 @@ namespace NexusLabs.Needlr.Generators.CodeGen;
 /// </summary>
 internal static class FactoryCodeGenerator
 {
+    private static string GetParameterBaseName(TypeDiscoveryHelper.ConstructorParameterInfo parameter)
+    {
+        if (!string.IsNullOrWhiteSpace(parameter.ParameterName))
+            return parameter.ParameterName!;
+
+        var typeName = GeneratorHelpers.GetShortTypeName(parameter.TypeName);
+        var genericStart = typeName.IndexOf('<');
+        if (genericStart >= 0)
+            typeName = typeName.Substring(0, genericStart);
+
+        return GeneratorHelpers.ToCamelCase(typeName);
+    }
+
+    private static string GetParameterIdentifierName(TypeDiscoveryHelper.ConstructorParameterInfo parameter)
+    {
+        return GeneratorHelpers.EscapeIdentifier(GetParameterBaseName(parameter));
+    }
+
+    private static Dictionary<string, string> GetInjectableNamesByType(
+        IEnumerable<TypeDiscoveryHelper.ConstructorParameterInfo> parameters)
+    {
+        var namesByType = new Dictionary<string, string>(StringComparer.Ordinal);
+        var usedNames = new HashSet<string>(StringComparer.Ordinal);
+
+        foreach (var parameter in parameters)
+        {
+            if (namesByType.ContainsKey(parameter.TypeName))
+                continue;
+
+            var baseName = GetParameterBaseName(parameter);
+            var uniqueName = baseName;
+            for (var suffix = 2; !usedNames.Add(uniqueName); suffix++)
+                uniqueName = baseName + suffix;
+
+            namesByType.Add(parameter.TypeName, uniqueName);
+        }
+
+        return namesByType;
+    }
+
     internal static void GenerateFactoryInterface(StringBuilder builder, DiscoveredFactory factory, BreadcrumbWriter breadcrumbs, string? projectDirectory)
     {
         var factoryName = $"I{factory.SimpleTypeName}Factory";
@@ -29,11 +69,8 @@ internal static class FactoryCodeGenerator
         // Generate Create method for each constructor
         foreach (var ctor in factory.Constructors)
         {
-            var runtimeParamList = string.Join(", ", ctor.RuntimeParameters.Select(p => 
-            {
-                var simpleTypeName = GeneratorHelpers.GetSimpleTypeName(p.TypeName);
-                return $"{p.TypeName} {p.ParameterName ?? GeneratorHelpers.ToCamelCase(simpleTypeName)}";
-            }));
+            var runtimeParamList = string.Join(", ", ctor.RuntimeParameters.Select(p =>
+                $"{p.TypeName} {GetParameterIdentifierName(p)}"));
 
             builder.AppendLine($"    /// <summary>Creates a new instance of {factory.SimpleTypeName}.</summary>");
             
@@ -42,7 +79,7 @@ internal static class FactoryCodeGenerator
             {
                 if (!string.IsNullOrWhiteSpace(param.DocumentationComment))
                 {
-                    var paramName = param.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(param.TypeName));
+                    var paramName = GetParameterBaseName(param);
                     var escapedDoc = GeneratorHelpers.EscapeXmlContent(param.DocumentationComment!);
                     builder.AppendLine($"    /// <param name=\"{paramName}\">{escapedDoc}</param>");
                 }
@@ -59,12 +96,12 @@ internal static class FactoryCodeGenerator
         var factoryInterfaceName = $"I{factory.SimpleTypeName}Factory";
         var factoryImplName = $"{factory.SimpleTypeName}Factory";
 
-        // Collect all unique injectable parameters across all constructors
         var allInjectableParams = factory.Constructors
             .SelectMany(c => c.InjectableParameters)
             .GroupBy(p => p.TypeName)
             .Select(g => g.First())
             .ToList();
+        var injectableNamesByType = GetInjectableNamesByType(allInjectableParams);
 
         builder.AppendLine("/// <summary>");
         builder.AppendLine($"/// Factory implementation for creating instances of <see cref=\"{factory.TypeName}\"/>.");
@@ -76,20 +113,24 @@ internal static class FactoryCodeGenerator
         // Fields for injectable dependencies
         foreach (var param in allInjectableParams)
         {
-            var fieldName = "_" + GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(param.TypeName));
+            var fieldName = "_" + injectableNamesByType[param.TypeName];
             builder.AppendLine($"    private readonly {param.TypeName} {fieldName};");
         }
 
         builder.AppendLine();
 
         // Constructor
-        var ctorParams = string.Join(", ", allInjectableParams.Select(p => $"{p.TypeName} {GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(p.TypeName))}"));
+        var ctorParams = string.Join(
+            ", ",
+            allInjectableParams.Select(p =>
+                $"{p.TypeName} {GeneratorHelpers.EscapeIdentifier(injectableNamesByType[p.TypeName])}"));
         builder.AppendLine($"    public {factoryImplName}({ctorParams})");
         builder.AppendLine("    {");
         foreach (var param in allInjectableParams)
         {
-            var fieldName = "_" + GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(param.TypeName));
-            var paramName = GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(param.TypeName));
+            var parameterName = injectableNamesByType[param.TypeName];
+            var fieldName = "_" + parameterName;
+            var paramName = GeneratorHelpers.EscapeIdentifier(parameterName);
             builder.AppendLine($"        {fieldName} = {paramName};");
         }
         builder.AppendLine("    }");
@@ -98,11 +139,8 @@ internal static class FactoryCodeGenerator
         // Create methods for each constructor
         foreach (var ctor in factory.Constructors)
         {
-            var runtimeParamList = string.Join(", ", ctor.RuntimeParameters.Select(p => 
-            {
-                var paramName = p.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(p.TypeName));
-                return $"{p.TypeName} {paramName}";
-            }));
+            var runtimeParamList = string.Join(", ", ctor.RuntimeParameters.Select(p =>
+                $"{p.TypeName} {GetParameterIdentifierName(p)}"));
 
             builder.AppendLine($"    public {factory.ReturnTypeName} Create({runtimeParamList})");
             builder.AppendLine("    {");
@@ -116,13 +154,13 @@ internal static class FactoryCodeGenerator
             var allArgs = new List<string>();
             foreach (var inj in ctor.InjectableParameters)
             {
-                var fieldName = "_" + GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(inj.TypeName));
-                var argName = inj.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(inj.TypeName));
+                var fieldName = "_" + injectableNamesByType[inj.TypeName];
+                var argName = GetParameterIdentifierName(inj);
                 allArgs.Add($"{argName}: {fieldName}");
             }
             foreach (var rt in ctor.RuntimeParameters)
             {
-                var paramName = rt.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(rt.TypeName));
+                var paramName = GetParameterIdentifierName(rt);
                 allArgs.Add($"{paramName}: {paramName}");
             }
 
@@ -141,8 +179,7 @@ internal static class FactoryCodeGenerator
         var funcType = $"Func<{runtimeTypes}, {factory.ReturnTypeName}>";
 
         // Build the lambda
-        var runtimeParams = string.Join(", ", ctor.RuntimeParameters.Select(p => 
-            p.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(p.TypeName))));
+        var runtimeParams = string.Join(", ", ctor.RuntimeParameters.Select(GetParameterIdentifierName));
 
         builder.AppendLine($"{indent}services.AddSingleton<{funcType}>(sp =>");
         builder.AppendLine($"{indent}    ({runtimeParams}) => new {factory.TypeName}(");
@@ -153,7 +190,7 @@ internal static class FactoryCodeGenerator
         var allArgs = new List<string>();
         foreach (var inj in ctor.InjectableParameters)
         {
-            var argName = inj.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(inj.TypeName));
+            var argName = GetParameterIdentifierName(inj);
             if (inj.IsKeyed)
             {
                 allArgs.Add($"{argName}: sp.GetRequiredKeyedService<{inj.TypeName}>(\"{GeneratorHelpers.EscapeStringLiteral(inj.ServiceKey!)}\")");
@@ -165,7 +202,7 @@ internal static class FactoryCodeGenerator
         }
         foreach (var rt in ctor.RuntimeParameters)
         {
-            var paramName = rt.ParameterName ?? GeneratorHelpers.ToCamelCase(GeneratorHelpers.GetSimpleTypeName(rt.TypeName));
+            var paramName = GetParameterIdentifierName(rt);
             allArgs.Add($"{paramName}: {paramName}");
         }
 
@@ -256,4 +293,3 @@ internal static class FactoryCodeGenerator
         return builder.ToString();
     }
 }
-

